src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs

```diff
@@ -573,12 +573,76 @@ private Task InvokeAllExceptionFiltersAsync()
 
                         _logger.ActionMethodExecuting(_actionExecutingContext, arguments);
 
-                        var actionReturnValue = await ControllerActionExecutor.ExecuteAsync(
-                            _executor,
-                            _controller,
-                            arguments);
+                        var returnType = _executor.MethodReturnType;
 
-                        result = CreateActionResult( actionMethodInfo.ReturnType, actionReturnValue);
+                        if (returnType == typeof(void))
+                        {
+                            _executor.Execute(_controller, arguments);
+                            result = new EmptyResult();
+                        }
+                        else if (returnType == typeof(Task))
+                        {
+                            await (Task)_executor.Execute(_controller, arguments);
+                            result = new EmptyResult();
+                        }
+                        else if (returnType == typeof(IActionResult))
+                        {
+                            result = (IActionResult)_executor.Execute(_controller, arguments);
+                            if (result == null)
+                            {
+                                throw new InvalidOperationException(
+                                    Resources.FormatActionResult_ActionReturnValueCannotBeNull(typeof(IActionResult)));
+                            }
+                        }
+                        else if (_executor.TaskGenericType == typeof(IActionResult))
+                        {
+                            result = await (Task<IActionResult>)_executor.Execute(_controller, arguments);
+                            if (result == null)
+                            {
+                                throw new InvalidOperationException(
+                                    Resources.FormatActionResult_ActionReturnValueCannotBeNull(typeof(IActionResult)));
+                            }
+                        }
+                        else if (_executor.IsTypeAssignableFromIActionResult)
```

Comment:
`_executor.TaskGenericType == typeof(IActionResult)` and `returnType == typeof(IActionResult)` - why are these cases handled separately from `_executor.IsTypeAssignableFromIActionResult`?